### PR TITLE
LPS-160293 Add type definition to FrontendDataSet component in frontend-data-set-web module

### PR DESCRIPTION
**Pull request by danvb112**
Related Issue: https://issues.liferay.com/browse/LPS-160293

### Motivation

Hey Reviewer, This PR follows the same idea as https://github.com/liferay-frontend/liferay-portal/pull/2608, but in this case I created the type definitions for the `FrontendDataSet` component.

 I'm working on a technical debt to remove some `@ts-ignore` annotations from the Objects and Notification code. For this I have to create some type declarations for components that were not developed in `Typescript`, I can declare these types in the `notification-web` module, but I think it will be restricted just for our case, so I decided to declare the types in the `frontend-data-set-web` which is the root of the `FrontendDataSet` component I had to remove the `@ts-ignore`, so anyone who wants to use this component with typescript in the future doesn't need to declare the same types in their own module, avoiding code repetition. What do you think?

### CODE REVIEW

**Comment by danvb112** [comment]
ci:test:relevant

File: modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+	};
+	inlineEditingSettings?: [
+		boolean,
+		{
+			alwaysOn: boolean;
+			defaultBodyContent: object;
+		}
+	];
+	items?: any[];
+	itemsActions?: any[];
```
Comment (diegonvs) [inline modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/index.d.ts:86]: Hey @danvb112, I think you can improve those types on itemsActions by taking a look on API table at https://liferay.atlassian.net/wiki/spaces/ENGFRONTENDINFRA/pages/2012217428/Frontend+Data+Set+Actions#Actions-Configuration

File: modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+		primaryItems?: any[];
+		secondaryItems?: any[];
+	};
+	currentURL?: string;
+	customDataRenderers?: any;
+	customViewsEnabled?: boolean;
+	enableInlineAddModeSetting?: {
+		defaultBodyContent?: object;
+	};
+	filters?: any[];
```
Comment (diegonvs) [inline modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/index.d.ts:69]: [Same](https://github.com/liferay-frontend/liferay-portal/pull/2627/files#r948029812) for filters, see https://liferay.atlassian.net/wiki/spaces/ENGFRONTENDINFRA/pages/2012184577/Frontend+Data+Set+Filters

**Comment by danvb112** [comment]
ci:test:sf

File: modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/index.d.ts (hunk 2 above)
Comment (danvb112) [reply modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/index.d.ts]: Hey @diegonvs, thanks for the suggestions :smile:, I don't think I really understand these types of `Custom Filters`, could you review what I did?

**Review by diegonvs — APPROVED** [review]

**Comment by danvb112** [comment]
ci:test:relevant

**Comment by diegonvs** [comment]
ci:forward

File: modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+type TFilters = {
+	id: string;
+	label: string;
+	type:
+		| 'autocomplete'
+		| 'checkbox'
+		| 'dateRange'
+		| 'number'
+		| 'radio'
+		| 'text';
```
Comment (markocikos) [inline modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/index.d.ts:157]: I updated [docs](https://liferay.atlassian.net/wiki/spaces/ENGFRONTENDINFRA/pages/2012184577/Frontend+Data+Set+Filters) on filters:
- The only valid ones are `autocomplete`, `dateRange` and `selection`. 
- `number` and `text` are fossils of some old implementations.
- We recently merged `checkbox` and `radio` into `selection`.
- I removed all docs for custom filters, they were wrong for sure. Custom filters currently don't work, although there is some code for them. We need to review the API, add a sample, and update docs after that. That's a seperate task. For this PR, we can safely remove all mentions of custom filters. 

You can confirm props in [the components code](https://github.com/liferay/liferay-portal/tree/master/modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/management_bar/components/filters). Note that at some point in the future, we plan to merge `autocomplete` into `selection` filter.

File: modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/index.d.ts [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+		confirmationMessage?: string;
+		id?: string;
+		method?: 'delete' | 'get';
+		permissionKey?: string;
+	};
+	href?: string;
+	icon?: string;
+	label?: string;
+	target?: 'async' | 'headless' | 'link' | 'modal' | 'sidePanel';
+};
```
Comment (markocikos) [inline modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/index.d.ts:146]: There seems to be an error here. When I run gradle deploy on `notification-web`, I get:
```
src/main/resources/META-INF/resources/js/components/DefinitionOfTerms.tsx(185,7): error TS2322: Type '{ id: string; items: Item[]; itemsActions: { href: string; id: string; label: string; target: string; }[]; namespace: string; onActionDropdownItemClick: <T>({ action, itemData, }: { action: FDSAction; itemData: T; }) => void; ... 6 more ...; views: { ...; }[]; }' is not assignable to type 'IFrontendDataSetProps'.
  Types of property 'itemsActions' are incompatible.
    Type '{ href: string; id: string; label: string; target: string; }[]' is not assignable to type 'TItemsActions[]'.
      Type '{ href: string; id: string; label: string; target: string; }' is not assignable to type 'TItemsActions'.
        Types of property 'target' are incompatible.
          Type 'string' is not assignable to type '"async" | "headless" | "link" | "modal" | "sidePanel" | undefined'.
```

**Review by markocikos — COMMENTED** [review]
@danvb112 I made a review on the closed PR, based on comments in https://github.com/brianchandotcom/liferay-portal/pull/121892. See comments inline.

Would it make sense to remove `propTypes`, along with adding TS types? It seems we are duplicating code with the same purpose. The only non-TS usage of `FrontendDataSet` in DXP in in the [FDS sample module](https://github.com/liferay/liferay-portal/tree/master/modules/apps/frontend-data-set/frontend-data-set-sample-web/src/main/resources/META-INF/resources/js), and we can easily migrate that to TS. That can be a part of this PR, or a separate task.

/cc @diegonvs

File: modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/index.d.ts (hunk 4 above)
Comment (diegonvs) [reply modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/index.d.ts:146]: Weird, I thought we can see some failures like this during a new `ant all` on CI or when `ant format-source-all` which is used for `ci:test:sf`

**Comment by diegonvs** [comment]
> Would it make sense to remove [these propTypes](https://github.com/liferay/liferay-portal/blob/master/modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/FrontendDataSet.js#L769-L838), along with adding TS types?

It makes sense 😅 . I completely forgot about this when reviewing. Hey @danvb112, could you update it?

File: modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/index.d.ts (hunk 3 above)
Comment (diegonvs) [reply modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/index.d.ts:157]: > number and text are fossils of some old implementations.

Jurassic Leftovers ®️ 😂 

 > We recently merged checkbox and radio into selection.

Oh, nice :)

**Pull request closed without merging**